### ISSUE TITLE
Basic interact check

### ISSUE THREAD

**Issue by SkirmisLTU** [GrimAnticheat/Grim#1391]
### The Suggestion
Hello, I would like to suggest adding an interaction check to Grim because currently, there is none. At the moment, the most needed check is to detect if a player is interacting with chests behind a wall or attacking another player behind a block (this is especially common for cheaters who use cheats to attack through cobwebs on certain PvP servers).

### Why It Fits Grim Anticheat?
I believe that it would be possible to create a check that is mathematically impossible to bypass since it would only need to trace where the player was looking, and if there are any blocks between the player who is interacting and the object that the player is trying to interact with. I think that a check like this would fit Grim's "philosophy" perfectly.

I know that this shouldn't be really hard for all the talented creators and contributors of Grim anticheat ;) Thanks.

**Comment by ManInMyVan** [comment]
So basically #1238 and #951?

**Comment by SkirmisLTU** [comment]
These suggestions aren't exactly the same as mine, but they're pretty similar. I see that there was some solutions, but I don't understand why they were abandoned and not implemented...

**Comment by SamB440** [comment]
#1238 is still a draft because noone else has been testing it and I don't like the implementation of handling uncertain block changes

**Comment by SkirmisLTU** [comment]
@SamB440, can I get the latest jar Grim with implementation of your code? I have medium sized server, where I can test it on my production environment and report if there is any issues with it. Thanks

**Comment by Symmettry** [comment]
Well, just make it cancel the interaction and don't ban them.

**Comment by SkirmisLTU** [comment]
> Well, just make it cancel the interaction and don't ban them.

I have no idea why are you saying me this, but I am still waiting for @SamB440 to get me the jar so I can begin testing

**Comment by SamB440** [comment]
Updated: https://github.com/SamB440/Grim/actions/runs/8308823011/artifacts/1332000061

**Comment by SkirmisLTU** [comment]
Hello again @SamB440, I've been using your version (with reach check for hitting trough walls) for more than a month now on my server (50-100 active players daily, PVP focused server). I ran into 0 problems. I think that it's safe to merge this update into main branch. Or maybe I can get update version of it?

**Comment by SamB440** [comment]
Done. https://github.com/SamB440/Grim/actions/runs/8815892204/artifacts/1442987269

**Comment by SkirmisLTU** [comment]
Hi one more time @SamB440,

I think it's time to merge this check. I've been using it for a long time now and haven't had any issues. My server has up to 100-200 players daily, so I believe that if there were any issues with this check, they would have shown by now. In case there is an issue with it, I suggest making a configuration option (such as `check-through-blocks: True`) so it can be disabled and not cause any problems.

If you're still not convinced, could I get an updated jar file?

**Comment by Symmettry** [comment]
> > Well, just make it cancel the interaction and don't ban them.
> 
> I have no idea why are you saying me this, but I am still waiting for @SamB440 to get me the jar so I can begin testing

uh i didn't reply to u, i commented on the thread. it probably notifies everyone

**Comment by SamB440** [comment]
Alright I shall clean it up and merge at some point

**Comment by SkirmisLTU** [comment]
Hi again @SamB440,

Do you have any news? I can't see the check being added anywhere. If you're not going to add it right now, could I ask you for a new jar with this check?

**Comment by SamB440** [comment]
I've been busy so haven't had much time, will do later

**Comment by SkirmisLTU** [comment]
Any updates?

**Comment by SamB440** [comment]
https://github.com/SamB440/Grim/actions/runs/9725696052/artifacts/1651545226

**Comment by SkirmisLTU** [comment]
Hi @SamB440, just a heads up that I'm still waiting for this to be merged to main branch;)

**Comment by SkirmisLTU** [comment]
Any news @SamB440 ?

**Comment by Axionize** [comment]
I've taken to maintaining a temporary fork with all the other latest GrimAC changes with Sam's reach and hitting through wall checks.

You can download the jars from releases here:
https://github.com/Axionize/Grim/releases

I'll be working on improving the performance of the checks. Please report any issues.

**Comment by SkirmisLTU** [comment]
Got no issues so far. I it would be nice if it could be updated or merged into main branch. @Axionize @SamB440

**Comment by SkirmisLTU** [comment]
Can I get an updated version, please? Since you don't want to implement it into Grim officially, I can continue testing and using it. Thanks, @SamB440!

**Comment by SamB440** [comment]
Use LightningGrim, axionize's fork @SkirmisLTU

**Comment by Axionize** [comment]
Yeah I've added a lot of changes from more reach check improvements, to performance improvements and inventory move, noslow, and more badpackets. The grim merge queue seems to move very slowly so I guess my fork is no longer temporary.

You can get the latest versions with all the updates, including mine and ones upstream from Grim here: https://modrinth.com/plugin/lightning-grim-anticheat

**Comment by SkirmisLTU** [comment]
Thank you for the suggestion. I tried using it, but it is too unstable for production environment unfortunately. Just walking and right clicking falses the aim, simulation and timer checks:(

**Comment by Axionize** [comment]
> Thank you for the suggestion. I tried using it, but it is too unstable for production environment unfortunately. Just walking and right clicking falses the aim, simulation and timer checks:(

My fork is used on several servers with hundreds of concurrent players individually, something's wrong in your setup... What plugins do you have? I think I would've noticed if walking falsed aim simulation and timer. Can I get a video?

The check works and Aim, Simulation, and Timer don't false from "just walking and right clicking". I can't reproduce this supposed "instability" issue.
https://streamable.com/hl9ww9

**Comment by ManInMyVan** [comment]
> Thank you for the suggestion. I tried using it, but it is too unstable for production environment unfortunately. Just walking and right clicking falses the aim, simulation and timer checks:(

disable `fix-1_21-placement-rotation` in your viaversion config

**Comment by SkirmisLTU** [comment]
Sorry for not replying earlier - I figured it out myself (my bad). However, at the moment, Lightning Grim is falling a bit behind on updates, unfortunately. I still believe the interact check should be merged, as I’ve had no issues with it for nearly a year on a server with 40+ players consistently. I think that’s enough to prove it’s stable enough. Thank you, @SamB440 ;)

**Comment by Axionize** [comment]
> Sorry for not replying earlier - I figured it out myself (my bad). However, at the moment, Lightning Grim is falling a bit behind on updates, unfortunately.

LightningGrim's beta builds are up to date with the latest changes from upstream, are you just looking at the releases perhaps?

> I still believe the interact check should be merged, as I’ve had no issues with it for nearly a year on a server with 40+ players consistently. I think that’s enough to prove it’s stable enough. Thank you, [@SamB440](https://github.com/SamB440) ;)

Unfortunately, this check is not quite at the mathematical rigor and correctness of Reach. We currently store a list of recently modified blocks to essentially exempt for and this allows bypasses and can false if more blocks than we maximally store are modified between postflys. Fixing this is possible but requires a lot more work.

Grim's first class checks (everything reach, hitbox) should be unbypassable and unfalseable logic checks. We will certainly merge the check in when it reaches that state but are unlikely to before then.
